runtime/near-vm-logic/src/gas_counter.rs

```diff
@@ -22,27 +22,98 @@ pub struct GasCounter {
     prepaid_gas: Gas,
     is_view: bool,
     ext_costs_config: ExtCostsConfig,
+
+    /// The cost of one regular operation in Gas.
+    regular_op_cost: Gas,
+    /// The amount regular operations a VM can do before it run out of gas (either max limit or
+    /// remaining prepaid gas).
+    remaining_regular_ops: u64,
+    /// The previous remaining amount of regular operations before `deduct_gas` was called.
+    /// By knowing the previous remaining amount and the current remaining amount, it's possible
+    /// to compute the number of operations used.
+    /// The gas counter is implemented this way to minimize the amount of logic needed for regular
+    /// gas counting.
+    prev_remaining_regular_ops: u64,
 }
 
 impl GasCounter {
     pub fn new(
+        regular_op_cost: Gas,
         ext_costs_config: ExtCostsConfig,
         max_gas_burnt: Gas,
         prepaid_gas: Gas,
         is_view: bool,
     ) -> Self {
-        Self { ext_costs_config, burnt_gas: 0, used_gas: 0, max_gas_burnt, prepaid_gas, is_view }
+        let mut counter = Self {
+            ext_costs_config,
+            burnt_gas: 0,
+            used_gas: 0,
+            max_gas_burnt,
+            prepaid_gas,
+            is_view,
+            regular_op_cost,
+            remaining_regular_ops: 0,
+            prev_remaining_regular_ops: 0,
+        };
+        counter.update_remaining_regular_ops();
+        counter
+    }
+
+    /// Recomputes `remaining_regular_ops` and `prev_remaining_regular_ops` based on the current
+    /// gas usage. It assumes the difference between `remaining_regular_ops` and
+    /// `prev_remaining_regular_ops` was already accounted in `burnt_gas` and `used_gas`.
+    fn update_remaining_regular_ops(&mut self) {
+        let remaining_burn_gas = if self.is_view {
+            self.max_gas_burnt - self.burnt_gas
+        } else {
+            std::cmp::min(self.max_gas_burnt - self.burnt_gas, self.prepaid_gas - self.used_gas)
+        };
+        // In case `regular_op_cost` is 0 when the fees are free, we don't want to divide by zero.
+        self.prev_remaining_regular_ops =
+            remaining_burn_gas.checked_div(self.regular_op_cost).unwrap_or(u64::MAX);
+        self.remaining_regular_ops = self.prev_remaining_regular_ops;
+    }
+
+    /// Optimized method to burn gas for the given number of regular operations.
+    pub fn burn_regular_ops(&mut self, num_regular_ops: u32) -> Result<()> {
```

Comment:
Why do you believe that this operation performs way better than original gas deduction? Cannot it be moved into contract instrumenting entirely? I.e. on contract startup we pass instructions budget and avoid host call completely, unless run out of budget? Although for that to work we need trusted location for the counter, that could only be decremented.

Reply:
Right and we also need to be able to update the counter whenever a host function is called. It's less secure than the current implementation. Also wasmer doesn't have large overhead for calling host functions